File: android/app/src/main/java/com/studyapp/domain/model/ProblemChapter.kt
package com.studyapp.domain.model

import java.util.UUID

data class ProblemChapter(
    val id: String = UUID.randomUUID().toString().lowercase(),
    val title: String = "章",
    val problemCount: Int = 0
)

data class ProblemNumberLocation(
    val globalNumber: Int,
    val chapterIndex: Int,
    val chapterTitle: String,
    val localNumber: Int
) {
    val displayText: String
        get() = "$chapterTitle ${localNumber}問"
}

fun List<ProblemChapter>.totalProblemCount(): Int =
    fold(0) { acc, chapter -> acc + maxOf(chapter.problemCount, 0) }

fun List<ProblemChapter>.locationFor(globalNumber: Int): ProblemNumberLocation? {
    if (globalNumber <= 0) return null
    var offset = 0
    for ((index, chapter) in this.withIndex()) {
        val count = maxOf(chapter.problemCount, 0)
        if (count <= 0) continue
        val rangeStart = offset + 1
        val rangeEnd = offset + count
        if (globalNumber in rangeStart..rangeEnd) {
            return ProblemNumberLocation(
                globalNumber = globalNumber,
                chapterIndex = index,
                chapterTitle = chapter.title,
                localNumber = globalNumber - offset
            )
        }
        offset += count
    }
    return null
}

fun List<ProblemChapter>.labelFor(globalNumber: Int): String =
    locationFor(globalNumber)?.displayText ?: "${globalNumber}問"

File: ios/StudyApp/Domain/Models/ProblemChapter.swift
import Foundation

struct ProblemChapter: Identifiable, Codable, Hashable {
    var id: String = UUID().uuidString.lowercased()
    var title: String
    var problemCount: Int

    private enum CodingKeys: String, CodingKey {
        case id
        case title
        case problemCount
    }

    init(id: String = UUID().uuidString.lowercased(), title: String, problemCount: Int) {
        self.id = id
        self.title = title
        self.problemCount = problemCount
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString.lowercased()
        title = try container.decodeIfPresent(String.self, forKey: .title) ?? "章"
        problemCount = try container.decodeIfPresent(Int.self, forKey: .problemCount) ?? 0
    }
}

struct ProblemNumberLocation: Hashable {
    var globalNumber: Int
    var chapterIndex: Int
    var chapterTitle: String
    var localNumber: Int

    var displayText: String {
        "\(chapterTitle) \(localNumber)問"
    }
}

extension Array where Element == ProblemChapter {
    var totalProblemCount: Int {
        reduce(0) { $0 + Swift.max($1.problemCount, 0) }
    }

    func location(for globalNumber: Int) -> ProblemNumberLocation? {
        guard globalNumber > 0 else { return nil }
        var offset = 0
        for (index, chapter) in enumerated() {
            let count = Swift.max(chapter.problemCount, 0)
            guard count > 0 else { continue }
            let range = (offset + 1)...(offset + count)
            if range.contains(globalNumber) {
                return ProblemNumberLocation(
                    globalNumber: globalNumber,
                    chapterIndex: index,
                    chapterTitle: chapter.title,
                    localNumber: globalNumber - offset
                )
            }
            offset += count
        }
        return nil
    }

    func label(for globalNumber: Int) -> String {
        location(for: globalNumber)?.displayText ?? "\(globalNumber)問"
    }
}
